### PR TITLE
MBS-11980: Test relationship types restriction in external links editor

### PR DESCRIPTION
**Pull request by reosarevok**
### Implement MBS-11980

### CODE REVIEW

**Comment by reosarevok** [comment]
This *should* now be correct - the new array features aren't really used by the test here, though, just by the one in https://github.com/metabrainz/musicbrainz-server/pull/2321

File: root/static/scripts/tests/Control/URLCleanup.js [hunk 1]
```diff
@@ -4984,6 +4989,55 @@ function doMatchSubtest(
...
+    }
+    return optionNames;
+  });
+
+  if (relationshipOptions.length > 0) {
+    actualRelationshipOptions = relationshipOptions;
+  }
+
+  const msg = 'Match ' + label + ' URL relationship options for ' +
+  entityType + ' entities';
```
Comment (mwiencek) [inline root/static/scripts/tests/Control/URLCleanup.js:5022]:
```suggestion
  const msg = 'Match ' + label + ' URL relationship options for ' +
    entityType + ' entities';
```

**Review by mwiencek — APPROVED** [review]

**Comment by reosarevok** [comment]
@yvanzo: given you've worked with this more, I'd still like another look from you before I merge :)

File: root/static/scripts/tests/Control/URLCleanup.js [hunk 2]
```diff
@@ -4529,14 +4529,18 @@ const testData = [
...
+ expected_relationship_options: ['downloadpurchase', 'streamingpaid'],
   },
   {
                      input_url: 'https://vimeo.com/ondemand/inconcert#comments',
             expected_clean_url: 'https://vimeo.com/ondemand/inconcert',
+             input_entity_type: 'recording',
        input_relationship_type: 'streamingpaid',
        only_valid_entity_types: ['recording', 'release'],
+ expected_relationship_options: ['downloadpurchase', 'streamingpaid'],
   },
```
Comment (yvanzo) [inline root/static/scripts/tests/Control/URLCleanup.js:4544]: These examples are wrong. The given URL `https://vimeo.com/ondemand/inconcert` should be linked using both `downloadpurchase`and `streamingpaid` relationship types. I don’t know if that applies to all Vimeo pages though. Sure things are:

* We need to know if all of Vimeo On Demand pages provide always both options or just any combination of these.
* Relationship types auto-selection for Vimeo On Demand URLs must be fixed to allow both relationship types at the same type (or to enforce selecting both if that matches Vimeo On Demand’s offer).
* Options example must be updated accordingly.

File: root/static/scripts/tests/Control/URLCleanup.js [hunk 3]
```diff
@@ -4984,6 +4989,55 @@ function doMatchSubtest(
   previousMatchTests.push(entityType + '+' + url);
 }
 
+function doOptionsSubtest(
```
Comment (yvanzo) [inline root/static/scripts/tests/Control/URLCleanup.js:4992]: I suggest naming tests of `CLEANUPS.restrict` with a more closely related name:

```suggestion
function doRestrictSubtest(
```

(“Options” may also refer to relationship attributes.)

File: root/static/scripts/tests/Control/URLCleanup.js [hunk 4]
```diff
@@ -4984,6 +4989,55 @@ function doMatchSubtest(
...
+) {
+  const checker = new Checker(cleanURL(url), entityType);
+  const possibleOptions = checker.getPossibleTypes();
+  let actualRelationshipOptions = possibleOptions || undefined;
+  if (expectedRelationshipOptions === undefined) {
+    actualRelationshipOptions = undefined;
+  }
+
+  // The options can be just one relationship type, or arrays of 2+ types
+  const relationshipOptions = possibleOptions.map(optionSet => {
```
Comment (yvanzo) [inline root/static/scripts/tests/Control/URLCleanup.js:5007]: I would suggest to consistenly use the term “combination” that is already used for this matter: `relationshipTypeCombinations`.

**Review by yvanzo — CHANGES_REQUESTED** [review]
The given example are not actually correct. It would be nice to have a better coverage of this new subtest with more diverse example allowing at least two different combinations and at least one combination of at least two relationship types.

File: root/static/scripts/tests/Control/URLCleanup.js (hunk 2 above)
Comment (reosarevok) [reply root/static/scripts/tests/Control/URLCleanup.js:4544]: It has to support all three options, as per the old discussion on https://tickets.metabrainz.org/browse/MBS-11952 - will fix accordingly.

**Review by yvanzo — APPROVED** [review]
Tested both tickets, that works nicely :) Made minor comments only. Also a more detailed commit message about the new `limited_link_type_combinations` test property would help me with writing a full documentation of URLCleanup tests from commit <https://github.com/metabrainz/musicbrainz-server/commit/dcabded48be065374c55bfb789ad588daa363729>’s message.

File: root/static/scripts/tests/Control/URLCleanup.js [hunk 5]
```diff
@@ -4984,6 +5003,55 @@ function doMatchSubtest(
...
+      expectedTypeCombinations.sort(),
+      (a, b) => {
+        if (Array.isArray(a) && Array.isArray(b)) {
+          return arraysEqual(a.sort(), b.sort());
+        } else if (typeof a === typeof b) {
+          return a === b;
+        }
+        return false;
+      },
+    ),
```
[the remark this answers is not in the document]
Comment (reosarevok) [reply root/static/scripts/tests/Control/URLCleanup.js:5049]: Good point, I think the sorts in the inner `arraysEqual` are no longer needed. Removing.

**Comment by reosarevok** [comment]
Updated the commit message with a lot more info, which is hopefully useful enough? :)

**Pull request merged**